src/server.c

```diff
@@ -97,49 +97,64 @@ const char *replstateToString(int replstate);
  * function of Redis may be called from other threads. */
 void nolocks_localtime(struct tm *tmp, time_t t, time_t tz, int dst);
 
+/* Close the currently open logfile and re-open it based on server.logfile */
+int serverLogOpen(void) {
+    if (server.log_fd != -1 && server.log_fd != STDOUT_FILENO) close(server.log_fd);
+    if (server.logfile[0] == '\0') {
+        server.log_fd = STDOUT_FILENO;
+    } else {
+        server.log_fd = open(server.logfile, O_APPEND|O_CREAT|O_WRONLY|O_CLOEXEC, 0644);
+    }
+    return server.log_fd == -1 ? C_ERR : C_OK;
+}
+
 /* Low level logging. To use only for very big messages, otherwise
  * serverLog() is to prefer. */
 void serverLogRaw(int level, const char *msg) {
     const int syslogLevelMap[] = { LOG_DEBUG, LOG_INFO, LOG_NOTICE, LOG_WARNING };
     const char *c = ".-*#";
-    FILE *fp;
-    char buf[64];
+    char buf[128];
     int rawmode = (level & LL_RAW);
-    int log_to_stdout = server.logfile[0] == '\0';
+    int ignored __attribute__((unused));
 
     level &= 0xff; /* clear flags */
     if (level < server.verbosity) return;
 
-    fp = log_to_stdout ? stdout : fopen(server.logfile,"a");
-    if (!fp) return;
-
+    if (server.syslog_enabled) syslog(syslogLevelMap[level], "%s", msg);
+    if (server.log_fd == -1 && serverLogOpen() != C_OK) return;
     if (rawmode) {
-        fprintf(fp,"%s",msg);
-    } else {
-        int off;
-        struct timeval tv;
-        int role_char;
-        pid_t pid = getpid();
+        ignored = write(server.log_fd,msg,strlen(msg));
+        return;
+    }
 
-        gettimeofday(&tv,NULL);
-        struct tm tm;
-        nolocks_localtime(&tm,tv.tv_sec,server.timezone,server.daylight_active);
-        off = strftime(buf,sizeof(buf),"%d %b %Y %H:%M:%S.",&tm);
-        snprintf(buf+off,sizeof(buf)-off,"%03d",(int)tv.tv_usec/1000);
-        if (server.sentinel_mode) {
-            role_char = 'X'; /* Sentinel. */
-        } else if (pid != server.pid) {
-            role_char = 'C'; /* RDB / AOF writing child. */
-        } else {
-            role_char = (server.masterhost ? 'S':'M'); /* Slave or Master. */
-        }
-        fprintf(fp,"%d:%c %s %c %s\n",
-            (int)getpid(),role_char, buf,c[level],msg);
+    pid_t pid = getpid();
```

Comment:
can we use server.pid & server.child_pid to instead of the `getpid`?before glibc2.24, getpid is still a costly syscall

Reply:
Not directly, because we need to address different cases involving Redis and/or module child processes. I don't think the additional optimization for very old glibc justifies the extra complexity.

Reply:
you're right，it probably won't give much performance improvement, and not worth the added complexity :-)